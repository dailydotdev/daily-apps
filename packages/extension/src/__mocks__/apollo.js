import { createMockClient } from 'mock-apollo-client';

// eslint-disable-next-line import/prefer-default-export
export const apolloClient = createMockClient();
