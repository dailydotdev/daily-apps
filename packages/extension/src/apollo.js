import { InMemoryCache } from 'apollo-cache-inmemory';
import { CachePersistor } from 'apollo-cache-persist';
import { createPersistedQueryLink } from 'apollo-link-persisted-queries';
import { createHttpLink } from 'apollo-link-http';
import { ApolloClient } from 'apollo-client';

const cache = new InMemoryCache();

export const persistor = new CachePersistor({
  cache,
  key: 'apollo-cache-persistor',
  storage: {
    async getItem(key) {
      const res = await browser.storage.local.get(key);
      if (!(Number.isNaN(res[key]) || res[key] === undefined)) {
        return res[key];
      }
      return undefined;
    },
    setItem: (key, data) => browser.storage.local.set({ [key]: data }),
    removeItem: key => browser.storage.local.remove(key),
  },
});

const link = createPersistedQueryLink({ useGETForHashedQueries: true })
  .concat(createHttpLink({
    uri: `${process.env.VUE_APP_API_URL}/graphql`,
    useGETForQueries: true,
    credentials: 'include',
  }));

export const apolloClient = new ApolloClient({
  link,
  cache,
});
