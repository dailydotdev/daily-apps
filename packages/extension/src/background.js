import { authService } from './common/services';
import {
  setCache,
  getCache,
  ANALYTICS_ID_KEY,
  DATA_VERSION_KEY,
  CURRENT_DATA_VERSION,
  STATE_KEY,
} from './common/cache';

const cacheUserId = async () => {
  const profile = await authService.getUserProfile();
  return setCache(ANALYTICS_ID_KEY, profile.id);
};

const migrateVersion1 = async () => {
  const [bookmarks, pubs, settings, profile, notificationTime] = await Promise.all([
    getCache('bookmarks', []),
    getCache('publications', {}),
    getCache('settings', {}),
    getCache('profile', null),
    getCache('notificationTime', null),
  ]);
  return {
    feed: {
      bookmarks,
      publications: Object.keys(pubs).map(id => ({ id, enabled: pubs[id] })),
    },
    ui: {
      lastNotificationTime: notificationTime,
      ...settings,
    },
    user: { profile },
  };
};

const migrateCache = async () => {
  const version = await getCache(DATA_VERSION_KEY, '1');
  if (version !== CURRENT_DATA_VERSION) {
    const newCache = await migrateVersion1();
    await setCache(STATE_KEY, newCache);
  }
  await setCache(DATA_VERSION_KEY, CURRENT_DATA_VERSION);
};

browser.browserAction.onClicked.addListener(() => {
  const url = browser.extension.getURL('index.html?source=button');
  browser.tabs.create({ url, active: true });
});

browser.alarms.create('cacheUserId', { periodInMinutes: 60 });

browser.alarms.onAlarm.addListener(async (alarm) => {
  if (alarm.name === 'cacheUserId') {
    // TODO: handle error
    await cacheUserId();
  }
});

browser.runtime.onInstalled.addListener(async () => {
  await Promise.all([
    cacheUserId(),
    migrateCache(),
  ]);
});
