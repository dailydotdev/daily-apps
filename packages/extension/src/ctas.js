import { browserName } from './common/browser';

const ctas = [{
  text: 'Daily is 100% open source',
  icon: 'github',
  link: 'https://github.com/dailydotdev/daily',
  name: 'Github',
  style: {
    color: 'white',
    background: 'black',
  },
}, {
  text: 'Follow us on Instagram',
  logo: 'instagram',
  link: 'https://www.instagram.com/dailydotdev/',
  name: 'Instagram',
  style: {
    color: '#838886',
    background: 'white',
  },
}, {
  text: 'Join our Slack',
  logo: 'slack',
  link: 'https://slack.daily.dev',
  name: 'Slack',
  style: {
    color: '#838886',
    background: 'white',
  },
}, {
  text: 'Follow real-time updates',
  icon: 'twitter',
  link: 'https://twitter.com/dailydotdev',
  name: 'Twitter',
  style: {
    color: 'white',
    background: '#00a2f9',
  },
}, {
  text: 'Like us on Facebook',
  icon: 'facebook',
  link: 'https://www.facebook.com/dailydotdev',
  name: 'Facebook',
  style: {
    color: 'white',
    background: '#3B5998',
  },
}];

if (browserName === 'chrome') {
  ctas.push({
    text: 'We need your review :)',
    logo: 'chrome_store',
    link: 'https://chrome.google.com/webstore/detail/daily-discover-web-techno/jlmpjdjjbgclbocgajdjefcidcncaied/reviews',
    name: 'Chrome',
    style: {
      color: '#838886',
      background: 'white',
    },
  });
} else if (browserName === 'firefox') {
  ctas.push({
    text: 'We need your review :)',
    logo: 'firefox',
    link: 'https://addons.mozilla.org/en-US/firefox/addon/daily/reviews/',
    name: 'Firefox',
    style: {
      color: 'white',
      background: '#0e1028',
    },
  });
}

export default ctas;
