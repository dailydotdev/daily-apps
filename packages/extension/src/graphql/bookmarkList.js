import gql from 'graphql-tag';

export const CREATE_BOOKMARK_LIST_MUTATION = gql`
mutation CreateBookmarkList($name: String!) {
  createBookmarkList(name: $name) {
    id, name
  }
}`;

export const RENAME_BOOKMARK_LIST_MUTATION = gql`
mutation RenameBookmarkList($id: ID!, $name: String!) {
  renameBookmarkList(id: $id, name: $name) {
    id, name
  }
}`;

export const REMOVE_BOOKMARK_LIST_MUTATION = gql`
mutation RemoveBookmarkList($id: ID!) {
  removeBookmarkList(id: $id) {
    _
  }
}`;

export const BOOKMARK_LISTS_QUERY = gql`
query BookmarkLists {
  bookmarkLists {
    id, name
  }
}`;
