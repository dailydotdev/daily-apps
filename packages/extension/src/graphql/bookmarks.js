import gql from 'graphql-tag';

export const ADD_BOOKMARKS_MUTATION = gql`
mutation AddBookmarks($data: AddBookmarkInput!) {
    addBookmarks(data: $data) {
        _
    }
}`;

export const ADD_BOOKMARK_TO_LIST_MUTATION = gql`
mutation AddBookmarkToList($id: ID!, $listId: ID) {
    addBookmarkToList(id: $id, listId: $listId) {
        _
    }
}`;

export const REMOVE_BOOKMARK_MUTATION = gql`
mutation RemoveBookmark($id: ID!) {
    removeBookmark(id: $id) {
        _
    }
}`;
