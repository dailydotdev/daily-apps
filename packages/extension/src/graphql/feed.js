import gql from 'graphql-tag';

export const FEED_POST_FRAGMENT = gql`
fragment FeedPost on Post {
  id,title,url,publishedAt,createdAt,image,ratio,placeholder,readTime,source { id, name, image },tags
}`;

export const USER_POST_FRAGMENT = gql`
fragment UserPost on Post {
  bookmarked,read
}`;

export const FEED_POST_CONNECTION_FRAGMENT = gql`
fragment FeedPostConnection on PostConnection {
  pageInfo { hasNextPage, endCursor }
  edges {
    node {
      ...FeedPost
      ...UserPost @include(if: $loggedIn)
    }
  }
}
${FEED_POST_FRAGMENT}
${USER_POST_FRAGMENT}`;

export const ANONYMOUS_FEED_QUERY = gql`
query AnonymousFeed($loggedIn: Boolean! = false, $now: DateTime!, $first: Int, $after: String, $ranking: Ranking, $filters: FiltersInput) {
  feed: anonymousFeed(now: $now, first: $first, after: $after, ranking: $ranking, filters: $filters) {
    ...FeedPostConnection
  }
}
${FEED_POST_CONNECTION_FRAGMENT}`;

export const FEED_QUERY = gql`
query Feed($loggedIn: Boolean! = false, $now: DateTime!, $first: Int, $after: String, $ranking: Ranking, $unreadOnly: Boolean) {
  feed(now: $now, first: $first, after: $after, ranking: $ranking, unreadOnly: $unreadOnly) {
    ...FeedPostConnection
  }
}
${FEED_POST_CONNECTION_FRAGMENT}`;

export const BOOKMARKS_FEED_QUERY = gql`
query BookmarksFeed($loggedIn: Boolean! = false, $now: DateTime!, $first: Int, $after: String, $unreadOnly: Boolean, $listId: ID) {
  feed: bookmarksFeed(now: $now, first: $first, after: $after, unreadOnly: $unreadOnly, listId: $listId) {
    ...FeedPostConnection
  }
}
${FEED_POST_CONNECTION_FRAGMENT}`;

export const SOURCE_FEED_QUERY = gql`
query SourceFeed($loggedIn: Boolean! = false, $now: DateTime!, $first: Int, $after: String, $ranking: Ranking, $source: ID!) {
  feed: sourceFeed(now: $now, first: $first, after: $after, ranking: $ranking, source: $source) {
    ...FeedPostConnection
  }
}
${FEED_POST_CONNECTION_FRAGMENT}`;

export const TAG_FEED_QUERY = gql`
query TagFeed($loggedIn: Boolean! = false, $now: DateTime!, $first: Int, $after: String, $ranking: Ranking, $tag: String!) {
  feed: tagFeed(now: $now, first: $first, after: $after, ranking: $ranking, tag: $tag) {
    ...FeedPostConnection
  }
}
${FEED_POST_CONNECTION_FRAGMENT}`;

export const SEARCH_POSTS_QUERY = gql`
query SearchPosts($loggedIn: Boolean! = false, $now: DateTime!, $first: Int, $after: String, $query: String!) {
  feed: searchPosts(now: $now, first: $first, after: $after, query: $query) {
    ...FeedPostConnection
    query
  }
}
${FEED_POST_CONNECTION_FRAGMENT}`;
