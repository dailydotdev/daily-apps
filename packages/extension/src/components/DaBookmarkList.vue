<template>
  <nav class="bookmark-list scrollbar">
    <div class="bookmark-list__container">
      <div class="bookmark-list__header nuggets">
        My bookmarks
      </div>
      <div class="bookmark-list__list">
        <button class="btn btn-menu" :class="{active: !bookmarkList}"
                @click="setBookmarkList(null)">
          <svgicon name="bookmark"/><span>All Articles</span>
        </button>
        <button class="btn btn-menu" :class="{active: bookmarkList === 'unread'}"
                @click="setBookmarkList('unread')">
          <svgicon name="unread"/><span>Unread</span>
        </button>
      </div>
    </div>
    <div class="bookmark-list__container">
      <div class="bookmark-list__header nuggets">
        <span>My lists</span>
        <button class="btn-icon btn-small" @click="openCreateList('Full')"
                v-tooltip="'Create new list'" v-if="bookmarkLists && bookmarkLists.length">
          <svgicon name="plus"/>
        </button>
      </div>
      <div class="bookmark-list__list">
        <button class="btn btn-menu bookmark-list__create-btn"
                v-if="!bookmarkLists || !bookmarkLists.length"
                @click="openCreateList('Empty')">
          <svgicon name="plus"/><span>Create new list</span>
        </button>
        <div v-for="item in bookmarkLists" :key="item.id" class="btn btn-menu"
            :class="{active: bookmarkList === item.id, hover: selectedListId === item.id}"
            @click="setBookmarkList(item.id)">
          <button class="bookmark-list__list-btn" @click="setBookmarkList(item.id)">
            {{item.name}}
          </button>
          <button class="btn-icon btn-small bookmark-list__menu-btn" @click="openMenu($event, item)"
                  :class="{active: selectedListId === item.id}">
            <svgicon name="menu"/>
          </button>
        </div>
      </div>
    </div>
    <da-context ref="context" class="bookmark-list__context" @open="onMenuOpened"
                @close="selectedList = null">
      <button class="btn btn-menu" @click="openRenameList">Rename</button>
      <button class="btn btn-menu" @click="openDeleteList">Delete</button>
    </da-context>
    <da-create-list v-if="showCreateList"
                    @close="showCreateList = false" />
    <da-rename-list v-if="showRenameList"
                    @close="showRenameList = false" :list="operationList" />
    <da-delete-list v-if="showDeleteList"
                    @close="showDeleteList = false" :list="operationList" />
  </nav>
</template>

<script>
import { mapState, mapGetters } from 'vuex';
import { BOOKMARK_LISTS_QUERY } from '../graphql/bookmarkList';

export default {
  name: 'DaBookmarkList',
  components: {
    DaContext: () => import('@daily/components/src/components/DaContext.vue'),
    DaCreateList: () => import('./DaCreateList.vue'),
    DaRenameList: () => import('./DaRenameList.vue'),
    DaDeleteList: () => import('./DaDeleteList.vue'),
  },

  apollo: {
    bookmarkLists: {
      query: BOOKMARK_LISTS_QUERY,
      fetchPolicy: 'cache-and-network',
      skip() {
        return !this.isPremium;
      },
    },
  },

  data() {
    return {
      bookmarkLists: [],
      selectedList: null,
      operationList: null,
      showCreateList: false,
      showRenameList: false,
      showDeleteList: false,
    };
  },

  computed: {
    selectedListId() {
      return this.selectedList && this.selectedList.id;
    },

    ...mapState('feed', ['bookmarkList']),
    ...mapGetters('user', ['isPremium']),
  },

  methods: {
    setBookmarkList(id) {
      let name;
      if (!id) {
        name = 'All';
      } else if (id === 'unread') {
        name = 'Unread';
      } else {
        name = 'Custom';
      }
      ga('send', 'event', 'Bookmark List', name);
      return this.$store.dispatch('feed/setBookmarkList', id);
    },
    openCreateList(mode) {
      if (this.isPremium) {
        ga('send', 'event', 'Bookmark List', 'Create', mode);
        this.showCreateList = true;
      } else {
        this.$store.commit('ui/setShowPremium', true);
      }
    },
    openMenu(event, list) {
      ga('send', 'event', 'Bookmark List', 'Context', 'Open');
      this.selectedList = list;
      this.$refs.context.open(event, list);
    },
    onMenuOpened(event) {
      const rect = event.target.getBoundingClientRect();
      this.$refs.context.positionMenu({ top: rect.bottom + 4, right: rect.right });
    },
    openRenameList() {
      ga('send', 'event', 'Bookmark List', 'Rename');
      this.operationList = this.selectedList;
      this.$refs.context.close();
      this.showRenameList = true;
    },
    openDeleteList() {
      ga('send', 'event', 'Bookmark List', 'Delete');
      this.operationList = this.selectedList;
      this.$refs.context.close();
      this.showDeleteList = true;
    },
  },

  mounted() {
    import('@daily/components/icons/bookmark');
    import('@daily/components/icons/unread');
    import('@daily/components/icons/plus');
    import('@daily/components/icons/menu');
  },
};
</script>

<style>
.bookmark-list {
  display: flex;
  flex-direction: column;
  overflow-y: auto;
  z-index: 2;
}

.bookmark-list__container {
  padding: 24px 0 8px;
  border-bottom: 1px solid var(--theme-background-highlight);

  &:nth-child(2) {
    border-bottom: none;
  }
}

.bookmark-list__header {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  margin: 0 16px 16px;
  color: var(--theme-disabled);
}

.bookmark-list__list {
  & .btn.btn-menu {
    width: 100%;
    height: 44px;
    padding: 10px 16px;
    --button-border-radius: 0 12px 12px 0;
    --button-color: var(--theme-secondary);
    @mixin micro1;
    text-transform: none;

    &:hover, &.hover {
      & .bookmark-list__menu-btn {
        visibility: visible;
      }
    }
  }
}

.bookmark-list__list-btn {
  flex: 1;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
  text-align: left;
  background: none;
  border: none;
  color: var(--button-color);
  cursor: pointer;
}

.bookmark-list__menu-btn {
  visibility: hidden;
}
</style>
