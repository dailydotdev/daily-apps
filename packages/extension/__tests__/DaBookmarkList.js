import { mount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import VueApollo from 'vue-apollo';
import icons from '@daily/components/src/icons';
import tooltip from '@daily/components/src/directives/tooltip';
import { apolloClient } from '../src/apollo';
import DaBookmarkList from '../src/components/DaBookmarkList.vue';
import { BOOKMARK_LISTS_QUERY } from '../src/graphql/bookmarkList';

jest.mock('../src/apollo');

let localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(VueApollo);
localVue.use(icons);
localVue.directive('tooltip', tooltip(localVue));

const bookmarkListHandler = jest.fn();
apolloClient.setRequestHandler(BOOKMARK_LISTS_QUERY, bookmarkListHandler);

let feed;
let user;
let store;

beforeEach(() => {
  window.ga = () => {
  };
  
  feed = {
    namespaced: true,
    state: {
      bookmarkList: null,
    },
    actions: {
      setBookmarkList: jest.fn(),
    },
  };

  user = {
    namespaced: true,
    state: {
      profile: { premium: true },
    },
    getters: {
      isLoggedIn: state => !!state.profile,
      isPremium: state => !!state.profile && !!state.profile.premium,
    },
  };

  store = new Vuex.Store({
    modules: { feed, user },
  });

  bookmarkListHandler.mockReset();
  bookmarkListHandler.mockResolvedValue({data: {bookmarkLists: 
    [{id: '1', name: 'handbook'}, {id: '2', name: 'favorite'}]
  }});
});

it('should set active the default list', () => {
  const wrapper = mount(DaBookmarkList, { store, localVue });
  expect(wrapper.findAll('.btn.btn-menu').at(0).classes()).toContain('active');
});

it('should set active the unread list', () => {
  feed.state.bookmarkList = 'unread';
  const wrapper = mount(DaBookmarkList, { store, localVue });
  expect(wrapper.findAll('.btn.btn-menu').at(1).classes()).toContain('active');
});

it('should set active the custom list', async () => {
  feed.state.bookmarkList = '2';
  const wrapper = mount(DaBookmarkList, { store, localVue, apolloProvider: new VueApollo({defaultClient: apolloClient }) });
  await wrapper.vm.$nextTick();
  expect(wrapper.findAll('.btn.btn-menu').at(3).classes()).toContain('active');
});

it('should dispatch setBookmarkList with no id', () => {
  const wrapper = mount(DaBookmarkList, { store, localVue });
  wrapper.findAll('.btn.btn-menu').at(0).trigger('click');
  expect(feed.actions.setBookmarkList).toBeCalledWith(expect.anything(), null);
});

it('should dispatch setBookmarkList with "unread" id', () => {
  const wrapper = mount(DaBookmarkList, { store, localVue });
  wrapper.findAll('.btn.btn-menu').at(1).trigger('click');
  expect(feed.actions.setBookmarkList).toBeCalledWith(expect.anything(), 'unread');
});

it('should dispatch setBookmarkList with the list id', async () => {
  const wrapper = mount(DaBookmarkList, { store, localVue, apolloProvider: new VueApollo({defaultClient: apolloClient }) });
  await wrapper.vm.$nextTick();
  wrapper.findAll('.bookmark-list__list-btn').at(1).trigger('click');
  expect(feed.actions.setBookmarkList).toBeCalledWith(expect.anything(), '2');
});

it('should show custom user lists', async () => {
  const wrapper = mount(DaBookmarkList, { store, localVue, apolloProvider: new VueApollo({defaultClient: apolloClient }) });
  await wrapper.vm.$nextTick();
  const lists = wrapper.findAll('.bookmark-list__list-btn');
  expect(lists.length).toEqual(2);
  const text = lists.wrappers.map(w => w.element.innerHTML);
  expect(text).toMatchSnapshot();
});

it('should open create list modal on plus button click', async () => {
  const wrapper = mount(DaBookmarkList, { store, localVue, apolloProvider: new VueApollo({defaultClient: apolloClient }) });
  expect(wrapper.find('.create-list').element).toBeFalsy();
  await wrapper.vm.$nextTick();
  wrapper.find('.bookmark-list__header button').trigger('click');
  await wrapper.vm.$nextTick();
  expect(wrapper.find('.create-list').element).toBeTruthy();
});

it('should open create list modal on dedicated button click', async () => {
  const wrapper = mount(DaBookmarkList, { store, localVue });
  expect(wrapper.find('.create-list').element).toBeFalsy();
  await wrapper.vm.$nextTick();
  wrapper.find('.bookmark-list__create-btn').trigger('click');
  await wrapper.vm.$nextTick();
  expect(wrapper.find('.create-list').element).toBeTruthy();
});

it('should open context menu', async () => {
  const wrapper = mount(DaBookmarkList, { store, localVue, apolloProvider: new VueApollo({defaultClient: apolloClient }) });
  expect(wrapper.find('.bookmark-list__context').element.style.display).toEqual('none');
  await wrapper.vm.$nextTick();
  wrapper.find('.bookmark-list__menu-btn').trigger('click');
  expect(wrapper.find('.bookmark-list__context').element.style.display).toEqual('none');
});

it('should open rename list modal', async () => {
  const wrapper = mount(DaBookmarkList, { store, localVue, apolloProvider: new VueApollo({defaultClient: apolloClient }) });
  expect(wrapper.find('.rename-list').element).toBeFalsy();
  await wrapper.vm.$nextTick();
  wrapper.find('.bookmark-list__menu-btn').trigger('click');
  wrapper.find('.bookmark-list__context button').trigger('click');
  await wrapper.vm.$nextTick();
  expect(wrapper.find('.rename-list').element).toBeTruthy();
});
