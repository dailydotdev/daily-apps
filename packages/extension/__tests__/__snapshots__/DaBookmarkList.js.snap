// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should show custom user lists 1`] = `
Array [
  "
          handbook
        ",
  "
          favorite
        ",
]
`;
